### PR TITLE
Update netlify adapter to integrate includeFiles and excludeFiles options

### PR DESCRIPTION
**Pull request by dfdez**
This PR implements the changes previously proposed in withastro/adapters#515, migrated to work within this repository's structure. The core functionality remains the same, with adjustments made to align with this repository.

## Changes
When trying Netlify SSR with Astro, I wasn't able to get the `included_files` option to work. After investigating, I realized that the `includeFiles` and `excludeFiles` functionalities, similar to what Vercel provides, were necessary to include additional files during the build process. 

In this PR, I have implemented the integration to add these options: 
- Added `includeFiles` and `excludeFiles` options to the integration, mirroring Vercel's functionality.
  - **`includeFiles`**: Allows users to explicitly specify files or directories to be bundled with their function. Useful for avoiding missing files in production.
  - **`excludeFiles`**: Enables users to specify files or directories to be excluded from the deployment. Helps minimize bundle size.
- All code for these features has been adapted directly from Vercel's integration for consistency and reliability.

## Testing

- Manually tested `includeFiles` and `excludeFiles` with various configurations to ensure:
  - Files specified in `includeFiles` are bundled correctly.
  - Files and directories specified in `excludeFiles` are omitted as expected.
- Automated tests added to validate edge cases and typical usage scenarios.
- Confirmed the implementation behaves identically to Vercel's integration.

## Docs

- Updated [astro docs](https://github.com/withastro/docs/pull/10760) with documentation for `includeFiles` and `excludeFiles` on netlify adapter.
- Added links to Vercel's documentation for further reference: 
  - [includeFiles](https://docs.astro.build/en/guides/integrations-guide/vercel/#includefiles)
  - [excludeFiles](https://docs.astro.build/en/guides/integrations-guide/vercel/#excludefiles)

### CODE REVIEW

File: packages/integrations/netlify/src/index.ts [hunk 1]
```diff
@@ -257,12 +278,38 @@ export default function netlifyIntegration(
...
 
+		const _includeFiles = integrationConfig?.includeFiles || [];
+		const _excludeFiles = integrationConfig?.excludeFiles || [];
+
+		if (finalBuildOutput === 'server') {
+			// Merge any includes from `vite.assetsInclude
+			if (_config.vite.assetsInclude) {
+				const mergeGlobbedIncludes = (globPattern: unknown) => {
+					if (typeof globPattern === 'string') {
+						const entries = glob.sync(globPattern).map((p) => pathToFileURL(p));
```
[the remark this answers is not in the document]
Comment (dfdez) [reply packages/integrations/netlify/src/index.ts:289]: > * Why aren't these assets resolved from the same `cwd` of line 263?

That is a good question; this code were taken from the vercel integration not sure why that difference does currently exists... 

Maybe it does have sense to update this and use the `getFilesByGlob` function to have the same behavior on either `includeFiles` and `vite.assetsInclude`.

> * Why don't we ignore `exclude`?

Similar to previous response; this code were taken from the vercel integration but I think there isn't any specific reason to don't `exclude` the files therefore we can add it.

I just wanted to keep the most consistency between integration, for those reasons I decided to keep this code as is. Let me know if any modifications are needed, and I'll gladly update it 👌

File: packages/integrations/netlify/src/index.ts [hunk 2]
```diff
@@ -139,6 +140,12 @@ async function writeNetlifyFrameworkConfig(config: AstroConfig, logger: AstroInt
 }
 
 export interface NetlifyIntegrationConfig {
+	/** Force files to be bundled with your function. This is helpful when you notice missing files. */
```
[the remark this answers is not in the document]
Comment (dfdez) [reply packages/integrations/netlify/src/index.ts]: I've updated the JS Docs based on your recommendations [here](https://github.com/withastro/astro/pull/13194/commits/93c2acfe1e52dd362960d6cabc69f9471eee9905). 

Thanks for the detailed feedback—please let me know if there are any further improvements you'd suggest.

File: packages/integrations/netlify/src/index.ts [hunk 3]
```diff
@@ -139,6 +140,12 @@ async function writeNetlifyFrameworkConfig(config: AstroConfig, logger: AstroInt
...
+	includeFiles?: string[];
+
+	/** Exclude any files from the bundling process that would otherwise be included. */
```
[the remark this answers is not in the document]
Comment (dfdez) [reply packages/integrations/netlify/src/index.ts]: Updated  [here](https://github.com/withastro/astro/pull/13194/commits/93c2acfe1e52dd362960d6cabc69f9471eee9905).

File: packages/integrations/netlify/src/index.ts [hunk 4]
```diff
@@ -246,6 +255,18 @@ export default function netlifyIntegration(
...
+	async function getFilesByGlob(
+		include: Array<string> = [],
+		exclude: Array<string> = [],
+	): Promise<Array<URL>> {
+		const files = await glob(include, {
+			cwd: fileURLToPath(rootDir),
+			absolute: true,
+			ignore: exclude,
+		});
+		return files.map((file) => new URL(pathToFileURL(file), _config.root));
```
[the remark this answers is not in the document]
Comment (dfdez) [reply packages/integrations/netlify/src/index.ts]: You are absolutely right. Updated [46a9895](https://github.com/withastro/astro/pull/13194/commits/46a9895037206e56747adde3cc4d895a81ff4cbe)

File: packages/integrations/netlify/src/index.ts [hunk 5]
```diff
@@ -139,6 +140,12 @@ async function writeNetlifyFrameworkConfig(config: AstroConfig, logger: AstroInt
 }
 
 export interface NetlifyIntegrationConfig {
+	/** Force files to be bundled with your function. This is helpful when you notice missing files. */
+	includeFiles?: string[];
+
+	/** Exclude any files from the bundling process that would otherwise be included. */
+	excludeFiles?: string[];
```
[the remark this answers is not in the document]
Comment (dfdez) [reply packages/integrations/netlify/src/index.ts:167]: I've added a note on the JS Docs https://github.com/withastro/astro/pull/13194/commits/93c2acfe1e52dd362960d6cabc69f9471eee9905#diff-c395613459736f27ff2bafd942481c8713261bd79538e1234994f9d3d4d0be5bR148

Also I will make sure to update the docs to add this important information 👍

File: packages/integrations/netlify/src/index.ts [hunk 6]
```diff
@@ -139,6 +140,32 @@ async function writeNetlifyFrameworkConfig(config: AstroConfig, logger: AstroInt
 }
 
 export interface NetlifyIntegrationConfig {
+	/**
+	 * Force files to be bundled with your SSR function.
+	 * This is useful for including any type of file that is not directly detected by the bundler,
+	 * like configuration files or assets that are dynamically imported at runtime.
+	 *
+	 * Note: File paths are resolved relative to your project's rootDir. Absolute paths may not work as expected.
```
Comment (ematipico) [inline packages/integrations/netlify/src/index.ts:148]:
```suggestion
	 * Note: File paths are resolved relative to your project's `rootDir`. Absolute paths may not work as expected.
```

File: packages/integrations/netlify/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply packages/integrations/netlify/src/index.ts:309]: It looks like the code is been there for a long time, let's keep it as is. Thank you for taking time for investigating and providing me with a proper response!

**Review by ematipico — APPROVED** [review]
Looks good to me!

File: packages/integrations/netlify/src/index.ts [hunk 7]
```diff
@@ -246,6 +275,18 @@ export default function netlifyIntegration(
...
+	async function getFilesByGlob(
+		include: Array<string> = [],
+		exclude: Array<string> = [],
+	): Promise<Array<URL>> {
+		const files = await glob(include, {
+			cwd: fileURLToPath(rootDir),
+			absolute: true,
+			ignore: exclude,
+		});
+		return files.map((file) => new URL(pathToFileURL(file)));
```
Comment (ascorbic) [inline packages/integrations/netlify/src/index.ts:287]:
```suggestion
		return files.map((file) => pathToFileURL(file));
```
`pathToFileURL` already return a URL object

File: .changeset/ninety-clouds-judge.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+---
+'@astrojs/netlify': minor
+---
+
+Adds `includedFiles` and `excludedFiles` configuration options to customize SSR function bundle contents.
+
+See the [documentation](https://docs.astro.build/en/guides/integrations-guide/netlify/) for detailed usage instructions and examples.
```
Comment (sarah11918) [inline .changeset/ninety-clouds-judge.md:7]:
``````suggestion

The `includeFiles` property allows you to explicitly specify additional files that should be bundled with your function. This is useful for files that aren't automatically detected as dependencies, such as:
- Data files loaded using `fs` operations
- Configuration files
- Template files

Similarly, you can use the `excludeFiles` property to prevent specific files from being bundled that would otherwise be included. This is helpful for:
- Reducing bundle size
- Excluding large binaries
- Preventing unwanted files from being deployed

```js
import { defineConfig } from 'astro/config';
import netlify from '@astrojs/netlify';

export default defineConfig({
  // ...
  output: 'server',
  adapter: netlify({
    includeFiles: ['./my-data.json'],
     excludeFiles: [
      './node_modules/package/**/*',
      './src/**/*.test.js'
    ],
  }),
});
```

See the [Netlify adapter documentation](https://docs.astro.build/en/guides/integrations-guide/netlify/#including-or-excluding-files) for detailed usage instructions and examples.
``````
Just a thought for making this a bit more hype, and providing content for the blog post! Correct/adjust as you like.

(Reminder: for the changeset, we're just alerting people to the feature and showing it off. We do NOT need caveats like "relative to the project root" or anything here. That's what the link to the documentation is for!)

**Review by sarah11918 — COMMENTED** [review]
Just left an idea of a sample changeset, for your consideration!

File: packages/integrations/netlify/src/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/integrations/netlify/src/index.ts:167]: Can you add it to the `excludeFiles` block too

File: packages/integrations/netlify/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/integrations/netlify/src/index.ts:309]: While the Vercel code has been there for a long time, this is new code for Netlify, so I think it would make sense to make it consistent with the cwd at least

**Review by ascorbic — APPROVED** [review]
